### DISCUSSION TITLE
Major difference in query likelihood retrieval performance when querying a pyserini index vs an indri index

### DISCUSSION THREAD

**Discussion by noamaon** [castorini/pyserini#1998] [Q&A] [answered]
Hi, I was comparing the performance of pyserini to the Indri toolkit and I got significant differences in performance. I compared NDCG@10, NDCG@5, MAP@50 and P@5 for a sample of 30 queries from the 2019 TREC retrieval challenge.
The corpus I was indexing is msmarco-passage (the first version). For both indices krovetz stemming was applied and the same mu parameter was set when retrieving. I even checked by hand that the text of the indexed passages is identical for both indices. Is this a known issue? 
Help will be much appreciated.

**Comment by lintool** [comment]
Yes, it is not a surprise that different retrieval engines produce slightly different results. How big are your differences in the various metrics?

**Reply by noamaon** [reply]
Thanks for the quick reply! The differences are not slight, attached is a picture containing the results I tested. Metrics for both Indri retrieval and Pyserini retrieval were calculated using the same script (simple parsing of the TREC output file together with the provided query relevance judgments from NIST). Are these results for Pyserini consistent with baselines you know?

![image](https://github.com/user-attachments/assets/822e3f56-5b88-4b2d-a544-e391f133db61)

**Reply by lintool** [reply]
Your pyserini results don't look right. Look here: https://castorini.github.io/pyserini/2cr/msmarco-v1-passage.html

To make sure you're doing everything right, I would suggest you go through the onboarding path here:
https://github.com/castorini/onboarding/blob/master/README.md#-foundations-of-retrieval

**Reply by noamaon** [reply]
Thanks again for the reply :) I will look into it. From a quick glance at the linked pages it seems there are only baselines for BM25 retrieval. To your knowledge, are QL retrieval results expected to be similar to the BM25 ones?

**Reply by lintool** [reply]
Yes, QL results should be similar to BM25, but why are you using QL? Most of the community uses BM25, and using QL will just lead to difficulty in performing fair comparisons...

**Reply by lintool** [reply]
But if you _really_ want QL, here's an example: https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-disk45.md

**Reply by noamaon** [reply]
Frankly, this seems quite similar to what I ran to achieve these results using pyserini.search.lucene with the 'qld' argument. I probably missed something in the process and am currently trying from scratch following the documentation. Are there any crucial arguments when building the index for QL retrieval?

**Reply by lintool** [reply]
The most obvious gotcha is tokenizer mismatch. You specify the tokenizer in indexing, and you have to use the same tokenizer during search.

**Reply by noamaon** [reply]
Aren't tokenizers only relevant with LM retrieval? How would a tokenizer be used for QL retrieval? I didn't specify a tokenizer when indexing...

**Reply by noamaon** [reply]
Additionally, I noticed that the qld scores I get from Pyserini are different than the ones when using Indri. The ones in the Pyserini output are all strictly positive while the ones in the Indri output are strictly negative. Indri uses either negative cross entropy or negative KL divergence as a similarity measure, What are the scores used by Pyserini? Is there some normalization process in the background?

**Reply by lintool** [reply]
- Tokenizers are applicable to Indri/Lucene as well - you've noted that you used krovetz stemming - that's tokenization.
- There are different ways to compute the score, but they are rank consistent - i.e., rank documents the same way.

**Reply by noamaon** [reply]
I see, I did apply stemming in the indexing process. is there a way to apply the same stemming in the querying process? This is probably not the only cause as I also tested this when not stemming and the differences were similar. Is it possible that some other stemmer is automatically applied in the querying process in Pyserini?

**Reply by lintool** [reply]
As I mentioned, your stemmer during indexing and querying have to align.

See https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/search/SearchCollection.java#L204

**Reply by noamaon** [reply]
Could you please point me towards the parameter which is relevant to the search stemming? I see multiple parameters containing analyzer/tokenizer and I'm not sure which one to use

**Reply by lintool** [reply]
Follow this for a complete worked about example: https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-disk45.md

**Reply by noamaon** [reply]
I must be missing something, but none of the examples there have a stemmer/tokenizer argument. Apologies for askig again, when I indexed the corpus I simply provided a "-stemmer " argument, is there a similar argument when querying?

**Reply by lintool** [reply]
On the indexing end, look here: https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/index/IndexCollection.java

In Lucene stemmers are called Analyzers.

Or you could just follow our guides and use our defaults... and not try to do anything "special" like messing with the stemmer, BM25 instead of QL, etc.

**Reply by noamaon** [reply]
I am using different parameters (QL, krovetz stemmer) for consistency purposes w.r.t other work from our lab. Is there anything stemming related on the querying end, or should I use the analyzer manually before performing the search? Additionally, even when using the default paramerers (default index stemmer, bm25) there are similar differences in performance when compared to Indri

**Reply by lintool** [reply, marked as answer]
I would try to reproduce our scores _exactly_ with default parameters (default index stemmer, bm25), then try to change one thing at a time... bm25 to QL, the stemmer, etc. and you'll find out where you went wrong.

**Reply by noamaon** [reply]
Hi, I found the issue - I think the search method automatically applies default stemming to queries unless stated otherwise, and regardless of the stemmer used for the corpus. Once I manually applied krovetz stemming to the queries and searched over the krovetz stemmed index, with the 'pretokenized' argument, everything worked as expected. Thank you very much for the quick replies, and for maintaining such a useful library!

**Reply by lintool** [reply]
Happy to help. Can you please close the loop by posting the updated results of Indri vs. Pyserini?

**Comment by noamaon** [comment]
Of course :) ![image](https://github.com/user-attachments/assets/3ee6406d-6f2f-4b2d-8571-07253ee2666a)

**Reply by lintool** [reply]
Okay, this looks as expected! Glad we resolved the issues.
